pallets/creators-staking/src/types.rs

```diff
@@ -0,0 +1,417 @@
+/// stake active again after 2 eras of inactivity.
+///
+/// **NOTE:** It is important to understand that staker **DID NOT** claim any rewards during this
+/// period.
+#[derive(Encode, Decode, Clone, Default, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct StakerInfo<Balance: AtLeast32BitUnsigned + Copy + MaxEncodedLen> {
+    // Size of this list would be limited by a configurable constant
+    stakes: Vec<EraStake<Balance>>,
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy + MaxEncodedLen> MaxEncodedLen for StakerInfo<Balance> {
+    // This is just an assumption, will be calculated properly in the future. See the comment for
+    // `MAX_ASSUMED_VEC_LEN`.
+    fn max_encoded_len() -> usize {
+        codec::Compact(MAX_ASSUMED_VEC_LEN).encoded_size().saturating_add(
+            (MAX_ASSUMED_VEC_LEN as usize).saturating_mul(EraStake::<Balance>::max_encoded_len()),
+        )
+    }
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy + MaxEncodedLen> StakerInfo<Balance> {
+    /// `true` if no active stakes and unclaimed eras exist, `false` otherwise
+    pub(super) fn is_empty(&self) -> bool {
+        self.stakes.is_empty()
+    }
+
+    /// number of `EraStake` chunks
+    pub(super) fn len(&self) -> u32 {
+        self.stakes.len() as u32
+    }
+
+    /// Stakes some value in the specified era.
+    ///
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    ///
+    /// # Example
+    ///
+    /// The following example demonstrates how internal vector changes when `stake` is called:
+    ///
+    /// `stakes: [<5, 1000>, <7, 1300>]`
+    /// * `stake(7, 100)` will result in `[<5, 1000>, <7, 1400>]`
+    /// * `stake(9, 200)` will result in `[<5, 1000>, <7, 1400>, <9, 1600>]`
+    pub(super) fn stake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if let Some(era_stake) = self.stakes.last_mut() {
+            if era_stake.era > current_era {
+                return Err("Unexpected era")
+            }
+
+            let new_stake_value = era_stake.staked.saturating_add(value);
+
+            if current_era == era_stake.era {
+                *era_stake = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes.push(EraStake::new(new_stake_value, current_era))
+            }
+        } else {
+            self.stakes.push(EraStake::new(value, current_era));
+        }
+
+        Ok(())
+    }
+
+    /// Unstakes some value in the specified era.
+    ///
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    ///
+    /// # Example 1
+    ///
+    /// `stakes: [<5, 1000>, <7, 1300>]`
+    /// * `unstake(7, 100)` will result in `[<5, 1000>, <7, 1200>]`
+    /// * `unstake(9, 400)` will result in `[<5, 1000>, <7, 1200>, <9, 800>]`
+    /// * `unstake(10, 800)` will result in `[<5, 1000>, <7, 1200>, <9, 800>, <10, 0>]`
```

Comment:
This code example is good.

_User should ensure that given era is either equal or greater than the latest available era in the staking info._
This means, we cannot modify eras in staking info, which are less than the latest.

This is the only rule, and we do not break it in the following example:
- On the 7th era we had 1200 total stake.
- We want to unstake 400 of those on the 9th era.
- We create **a new** record for the 9th era, which says that our stake had decreased.

When the rewards will be calculated, it will consider that we had staked 1200 for the 7th and 8th era, and 800 starting from 9th era.

Reply:
This code example is good.

_User should ensure that given era is either equal or greater than the latest available era in the staking info._
This means, we cannot modify eras in staking info, which are less than the latest.

That's the only rule, and we don't This code example works the next way:
- On the 7th era we had 1200 total stake.
- We want to say, that we unstake 400 of those on the 9th era.
- We create **a new** record for the 9th era, which says that our stake had decreased.